File: src/main/java/cn/keevol/keenotes/mobilefx/DebugView.java
package cn.keevol.keenotes.mobilefx;

import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.scene.layout.Priority;

/**
 * Debug view for development and troubleshooting.
 * Contains all debug-related functionality.
 */
public class DebugView extends VBox {

    private final Label statusLabel;

    public DebugView(Runnable onBack) {
        getStyleClass().add("debug-view");

        // Status label for feedback
        statusLabel = new Label();
        statusLabel.getStyleClass().add("debug-status");
        statusLabel.setWrapText(true);

        // Debug buttons container
        VBox debugContainer = new VBox(12);
        debugContainer.setPadding(new Insets(16));
        debugContainer.getStyleClass().add("debug-container");

        // Add status label and all debug buttons (matching Android order)
        debugContainer.getChildren().addAll(
            statusLabel,
            createDebugButton("Check DB Count", this::checkDbCount),
            createDebugButton("Dump All Notes", this::dumpAllNotes),
            createDebugButton("Reset Sync State", this::resetSyncState),
            createDebugButton("Clear All Notes", this::clearAllNotes),
            createDebugButton("Test WebSocket", this::testWebSocket)
        );

        // Spacer to push content up
        VBox.setVgrow(debugContainer, Priority.ALWAYS);

        // Header with back button and debug container
        getChildren().addAll(createHeader(onBack), debugContainer);
    }

    private HBox createHeader(Runnable onBack) {
        Label title = new Label("Debug View");
        title.getStyleClass().add("debug-title");

        Button backBtn = new Button("Back");
        backBtn.getStyleClass().add("back-btn");
        backBtn.setOnAction(e -> onBack.run());

        HBox header = new HBox(8, title, backBtn);
        header.setAlignment(Pos.CENTER_LEFT);
        header.setPadding(new Insets(12, 16, 8, 16));
        header.getStyleClass().add("header");
        return header;
    }

    private Button createDebugButton(String text, Runnable action) {
        Button btn = new Button(text);
        btn.getStyleClass().add("debug-button");
        btn.setMaxWidth(Double.MAX_VALUE);
        btn.setOnAction(e -> action.run());
        return btn;
    }

    private void dumpAllNotes() {
        LocalCacheService localCache = ServiceManager.getInstance().getLocalCacheService();
        if (localCache == null || !localCache.isInitialized()) {
            updateStatus("LocalCache not ready");
            return;
        }
        var notes = localCache.getAllNotes();
        StringBuilder sb = new StringBuilder("=== Recent Notes ===\n");
        int count = 0;
        for (var note : notes) {
            if (count++ >= 20) break;
            String preview = note.content != null ? note.content.substring(0, Math.min(50, note.content.length())) : "";
            sb.append("ID: ").append(note.id).append(", Content: ").append(preview).append("...\n");
        }
        updateStatus(sb.toString());
        System.out.println(sb);
    }

    private void resetSyncState() {
        LocalCacheService localCache = ServiceManager.getInstance().getLocalCacheService();
        if (localCache == null || !localCache.isInitialized()) {
            updateStatus("LocalCache not ready");
            return;
        }
        localCache.resetSyncState();
        updateStatus("Sync state reset to initial");
        System.out.println("=== SYNC STATE RESET ===");
    }

    private void checkDbCount() {
        LocalCacheService localCache = ServiceManager.getInstance().getLocalCacheService();
        if (localCache == null || !localCache.isInitialized()) {
            updateStatus("LocalCache not ready");
            return;
        }
        int count = localCache.getLocalNoteCount();
        long lastSyncId = localCache.getLastSyncId();
        String lastSync = localCache.getLastSyncTime();
        String msg = "DB has " + count + " notes\nLast sync ID: " + lastSyncId + "\nLast sync time: " + (lastSync != null ? lastSync : "Never");
        updateStatus(msg);
        System.out.println("=== DEBUG: " + msg + " ===");
    }

    private void clearAllNotes() {
        LocalCacheService localCache = ServiceManager.getInstance().getLocalCacheService();
        if (localCache == null || !localCache.isInitialized()) {
            updateStatus("LocalCache not ready");
            return;
        }
        localCache.clearAllData();
        updateStatus("All notes cleared");
        System.out.println("=== ALL NOTES CLEARED ===");
    }

    private void testWebSocket() {
        WebSocketClientService wsService = ServiceManager.getInstance().getWebSocketService();
        if (wsService == null) {
            updateStatus("WebSocket service not available");
            return;
        }
        
        boolean connected = wsService.isConnected();
        updateStatus("WebSocket state: " + (connected ? "CONNECTED" : "DISCONNECTED"));
        
        if (!connected) {
            wsService.connect();
            updateStatus("Attempting to connect...");
        } else {
            // Force reconnect to trigger sync
            wsService.disconnect();
            updateStatus("Disconnected, reconnecting in 1s...");
            new Thread(() -> {
                try {
                    Thread.sleep(1000);
                    wsService.connect();
                    javafx.application.Platform.runLater(() -> updateStatus("Reconnecting..."));
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }).start();
        }
    }

    private void updateStatus(String message) {
        javafx.application.Platform.runLater(() -> statusLabel.setText(message));
    }
}

File: keenotes-ios/KeeNotes/Views/DebugView.swift
import SwiftUI

/// Debug view for development and troubleshooting
struct DebugView: View {
    @EnvironmentObject var appState: AppState
    @Environment(\.dismiss) var dismiss
    
    @State private var statusText = ""
    
    var body: some View {
        NavigationView {
            VStack(alignment: .leading, spacing: 16) {
                // Status text
                if !statusText.isEmpty {
                    Text(statusText)
                        .font(.system(.caption, design: .monospaced))
                        .foregroundColor(.secondary)
                        .padding()
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(Color(.systemGray6))
                        .cornerRadius(8)
                }
                
                // Debug buttons
                VStack(spacing: 12) {
                    DebugButton(title: "Check DB Count", action: checkDbCount)
                    DebugButton(title: "Dump All Notes", action: dumpAllNotes)
                    DebugButton(title: "Reset Sync State", action: resetSyncState)
                    DebugButton(title: "Clear All Notes", action: clearAllNotes)
                    DebugButton(title: "Test WebSocket", action: testWebSocket)
                }
                
                Spacer()
            }
            .padding()
            .navigationTitle("Debug View")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    Button("Back") {
                        dismiss()
                    }
                }
            }
        }
    }
    
    private func checkDbCount() {
        Task {
            do {
                let count = try await appState.databaseService.getNoteCount()
                let syncState = try await appState.databaseService.getSyncState()
                let msg = """
                DB has \(count) notes
                Last sync ID: \(syncState?.lastSyncId ?? -1)
                Last sync time: \(syncState?.lastSyncTime ?? "Never")
                """
                await MainActor.run {
                    statusText = msg
                }
                print("[Debug] \(msg)")
            } catch {
                await MainActor.run {
                    statusText = "Error: \(error.localizedDescription)"
                }
            }
        }
    }
    
    private func dumpAllNotes() {
        Task {
            do {
                let notes = try await appState.databaseService.getRecentNotes(limit: 20)
                var sb = "=== Recent 20 Notes ===\n"
                for note in notes {
                    let preview = String(note.content.prefix(50))
                    sb += "ID: \(note.id), Content: \(preview)...\n"
                }
                await MainActor.run {
                    statusText = sb
                }
                print("[Debug] \(sb)")
            } catch {
                await MainActor.run {
                    statusText = "Error: \(error.localizedDescription)"
                }
            }
        }
    }
    
    private func resetSyncState() {
        Task {
            do {
                try await appState.databaseService.clearSyncState()
                await MainActor.run {
                    statusText = "Sync state reset, reconnecting..."
                }
                print("[Debug] === SYNC STATE RESET ===")
                
                appState.webSocketService.disconnect()
                try? await Task.sleep(nanoseconds: 500_000_000)
                appState.webSocketService.connect()
                
                await MainActor.run {
                    statusText = "Sync state reset, reconnected"
                }
            } catch {
                await MainActor.run {
                    statusText = "Error: \(error.localizedDescription)"
                }
            }
        }
    }
    
    private func clearAllNotes() {
        Task {
            do {
                try await appState.databaseService.clearAllData()
                await MainActor.run {
                    statusText = "All notes cleared, reconnecting..."
                }
                print("[Debug] === ALL NOTES CLEARED ===")
                
                appState.webSocketService.disconnect()
                try? await Task.sleep(nanoseconds: 500_000_000)
                appState.webSocketService.connect()
                
                await MainActor.run {
                    statusText = "All notes cleared, reconnected"
                }
            } catch {
                await MainActor.run {
                    statusText = "Error: \(error.localizedDescription)"
                }
            }
        }
    }
    
    private func testWebSocket() {
        let state = appState.webSocketService.connectionState
        statusText = "WebSocket state: \(state)"
        
        switch state {
        case .disconnected:
            appState.webSocketService.connect()
            statusText = "Attempting to connect..."
        case .connected:
            appState.webSocketService.disconnect()
            statusText = "Disconnected, reconnecting in 1s..."
            Task {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                appState.webSocketService.connect()
                await MainActor.run {
                    statusText = "Reconnecting..."
                }
            }
        case .connecting:
            statusText = "Already connecting..."
        }
    }
}

/// Styled debug button
struct DebugButton: View {
    let title: String
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            Text(title)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 12)
                .background(Color.blue)
                .foregroundColor(.white)
                .cornerRadius(8)
        }
    }
}
